### PR TITLE
[5.8] Add @error blade directive

### PR DESCRIPTION
**Pull request by calebporzio**
This PR comes out of a popular tweet I wrote: https://twitter.com/calebporzio/status/1111366190444802048

```php
// Before
@if ($errors->has('email'))
    <span>{{ $errors->first('email') }}</span>
@endif
  
// After:
@error('email')
    <span>{{ $message }}</span>
@enderror
```

### CODE REVIEW

**Review by driesvints — APPROVED** [review]

**Comment by driesvints** [comment]
Pretty neat!

**Comment by browner12** [comment]
minor suggestion, switch to `CompilesErrors` (plural) so it stays consistent with other traits?

**Comment by driesvints** [comment]
@calebporzio I think the suggestion by @browner12 is valid. Can you update the PR?

**Comment by devcircus** [comment]
Sounds like a good start to adding blade helpers to your awesome-helpers package, or an awesome-blade-helpers package.

**Comment by jbrooksuk** [comment]
FWIW, this was previously closed. https://github.com/laravel/framework/pull/20111

**Comment by rellect** [comment]
Problem with such helpers is that they are not flexible, they handle a very specific case. For example, you can't check if error does not exists, so in this case and many others you'll have to fallback to a regular if statement. Personally I try to keep my code consistent so I avoid using such helpers.

**Comment by calebporzio** [comment]
Updated - thanks!

**Comment by mo7amed-3bdalla7** [comment]
I made a PR before like this and closed #27944

**Comment by driesvints** [comment]
@mo7amed-3bdalla7 this is a different implementation.

**Comment by garygreen** [comment]
Problem I have with stuff like this is although on the face of it it looks nice - your still find yourself reaching for `$errors->has('email')` - often you want to add a "has-error" class to the container element, plus render a message:

```html
<div class="form-group {{ $errors->has('email') ? 'has-error' : '' }}">
	<input type="email">
	@error('email')
		<span class="help-block">{{ $message }}</span>
	@enderror
</div>
```

...developers may get confused (especially newcomers) as to why they need to use both `$errors` and `@error` in their view.

Rendering errors is always something I've found a bit clumsy, so in previous projects I've experimented using a "wrapper" around the container for the field, something along these lines:

```html
@field('email', ['class' => 'flex']) <!-- <div class="flex form-group has-error"> -->
	<input type="email">
	{{ $error }} <!-- <span class="help-block">...</span> -->
@endField
```

...but even that style I've started to go off because, although it reduces amount of code, it's a bit cryptic in what it's actually going to render. It also creates a disparity between how you render errors in js e.g. Vue/React, to how you do it in blade. Purely for explicitness, I've started just writing it out in full, that way developers know exactly what it's going to render, how to render errors, why it's rendering that way etc.

In summary, I've still not come up with a solution I've personally been that comfortable using in my apps, but this PR looks interesting as a way of solving the "inner" message problem but doesn't solve the "outer" problem, if that makes sense. I personally would say for this reason it's maybe something best left in userland.

**Comment by calebporzio** [comment]
Makes sense, what about this:

```
<div class="form-group @error('email', 'has-error')">
	<input type="email">
	@error('email')
		<span class="help-block">{{ $message }}</span>
	@enderror
</div>
```

I kinda dig that. There could be a third parameter default too: `<input class="@error('email', 'border-red', 'border-gray')">`

I get just wanting to keep the original `$errors`, but damn it's a lot of ugly code for such a common use case. If it doesn't feel ugly to you, look no further than the likes and comments on the tweet.

**Comment by m1guelpf** [comment]
For reference: Here's my #noplanstomerge PR to 5.4 https://github.com/laravel/framework/pull/20111

**Comment by calebporzio** [comment]
Ok, just actually needed to set an "input error" state on an input element. I much prefer this syntax to the standard blade echo ternary:

Before:
image

After:
image

**Comment by ankurk91** [comment]
Does it handle named error bags? Something like?

```php
Blade::directive('isInvalid', function ($expression) {
            $segments = explode(',', $expression);
            $input = trim($segments[0]);
            $errorBag = trim(Arr::get($segments, 1, 'default'), "'");

            return '<?php echo $errors->'.$errorBag.'->has('.$input.') ? "is-invalid" : ""; ?>';
        });
```

Why not two directives?
```php
// Bootstrap css classes
Blade::directive('validationMessage', function ($expression) {
            $segments = explode(',', $expression);
            $input = trim($segments[0]);
            $errorBag = trim(Arr::get($segments, 1, 'default'), "'");

            return '<?php if ($errors->'.$errorBag.'->has('.$input.')) { ?>
                      <div class="invalid-feedback d-block" role="alert">
                        <?php echo e($errors->'.$errorBag.'->first('.$input.')); ?>
                      </div>
                  <?php } ?>';
        });
```

**Comment by calebporzio** [comment]
@ankurk91 - the goal of this feature is to cover (what I believe to be) the 90% of use cases for `$errors`. I think named error bags and other features shouldn't be handled by it. It's simplicity will start to erode.

**Comment by garygreen** [comment]
I'm still not convinced this is something that should be added to core. There are so many opinionated ways you could do error handling in views, getting involved in this logic is probably a recipe for disaster down the line.

### Style 1 - render second parameter if contains an error
```blade
<div class="form-group @error('email', 'has-error')">
	<input type="email">
	@error('email')
		<span class="help-block">{{ $message }}</span>
	@enderror
</div>
```

### Style 2 - allow html as second parameter, to format error
```blade
<div class="form-group @error('email', 'has-error')">
	<input type="email">
	@error('email', '<span class="help-block">:message</span>')
</div>
```

### Style 3 - using a custom helpers to return class/html formatted errors:
```blade
<div class="form-group @errorClass('email')">
	<input type="email">
	@errorHtml('email') <!-- outputs <span class="help-block">.... if there is an error !-->
</div>
```

### Style 4 - using short Emmet-style syntax to reduce amount of typing of second parameter
```blade
<div class="form-group @error('email', 'has-error')">
	<input type="email">
	@error('email', 'span.help-block') <!-- will be auto-closed and contain :message -->
</div>
```

All of the above to avoid:

### Current:

```blade
<div class="form-group {{ $errors->has('email') ? 'has-error' : '' }}">
	<input type="email">
	@if ($errors->has('email'))
		<span class="help-block">{{ $errors->first('email') }}</span>
	@endif
</div>
```

**Comment by calebporzio** [comment]
"All of the above to avoid" I don't think the mentioned "above" is a valid problem statement. I never proposed offering Emmet style syntax.

Here is what I'm proposing:

```php
@error('first_name')
    <span>{{ $message }}</span>
@enderror
```
to avoid having to write
```php
@if ($errors->has('first_name'))
    <span>{{ $errors->first('first_name') }}</span>
@endif
```

This will cover most use cases. In the case of needing an inline style, you can do this:

```php
<input name="first_name" class="@error('first_name') has-error @enderror">

@error('first_name)
    {{ $message }}
@enderror
```

I think this is cleaner. I think this is prettier. A bunch of other people think so too.

**Comment by efrain-salas** [comment]
What about that ?

https://twitter.com/michaeldyrynda/status/1115458350710321153?s=21

**Comment by garygreen** [comment]
> I think this is cleaner. I think this is prettier. A bunch of other people think so too.

I'm not disagreeing behind the intentions of this PR - reducing the amount of code to write is always a good developer experience, but I personally think we should be careful adding new blade stuff like this without properly understanding why it's needed.

This PR introduces side effects, by which I mean: what happens in the current PR you `@error()` into multiple blocks? What if you forget to `@enderror`? Can I still use `$message` and  `$messageCache` in my views? What about current Laravel tutorial/guides that show all about `$errors` and now there's a new way to do it with `@error`, would that be helpful or confusing to newcomers?

Most will solve the problem in userland by writing their own helpers which can fully format the error message to whatever framework format they are using without side effects, or use IDE helpers to write their own keyboard shortcut snippets.

For example, you can write some very basic directives to do this kind of format, which IMO is even simpler and more declarative in usage than this PR.

```blade
<div class="form-group @hasError('email')">
  <input type="email" class="form-control">
	@error('email') <!-- <span class="help-block">:message</span> -->
</div>
```

```php
\Blade::directive('hasError', function($expression) {
	return "<?= \$errors->has($expression) ? 'has-error' : '' ?>";
});

\Blade::directive('error', function($expression) {
	return "<?= \$errors->first($expression, '<span class=\"help-block\">:message</span>') ?>";
});
```

Again, I'm not totally against this PR, but I just question it's usefulness and appeal in real world applications when there are alternatives in userland.

File: src/Illuminate/View/Compilers/Concerns/CompilesErrors.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    /**
+     * Compile the enderror statements into valid PHP.
+     *
+     * @param  string  $expression
+     * @return string
+     */
+    protected function compileEnderror($expression)
+    {
+        return '<?php unset($message);
+if (isset($messageCache)) { $message = $messageCache; }
```
[the remark this answers is not in the document]
Comment (garygreen) [reply src/Illuminate/View/Compilers/Concerns/CompilesErrors.php]: Why would you need to nest? You can use dot format to render deeply nested error messages.

[the remark this answers is not in the document]
Comment (calebporzio) [reply src/Illuminate/View/Compilers/Concerns/CompilesErrors.php]: Sorry, I said "believe" not to be confrontational, but to express I don't actually know for sure. I meant "believe" to be a loose word, not strong lol.

I'm pretty indifferent on supporting nesting because I can't really see myself using it. But I agree it seems like the behavior is unexpected if what you're saying is true.

Can you provide the solution?

**Comment by calebporzio** [comment]
Agreed that this can be accomplished debatably better in userland, but that goes for most syntax like this.

I would say that this is such a rediculously common use case, it makes sense to make it nicer. It's like `@csrf` or `@old`. They provide pleasant syntax for common use cases. 

Good point re onboarding newcomers. I think this could be the primary way of documenting and then, an example/note using the more verbose `$errors` to show it's available. I think this could be the idiomatic way to access validation errors and it would make onboarding even easier.

I believe it's useful for most applications using Blade.

File: src/Illuminate/View/Compilers/Concerns/CompilesErrors.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (d3jn) [reply src/Illuminate/View/Compilers/Concerns/CompilesErrors.php]: > Can you provide the solution?

Just utilize `array_pop`/`array_push` and make `$messageCache` into array instead. This will look something like this:

```php
    // For opening directive
    return '<?php if ($errors->has('.$expression.')) :
        if (! isset($message)) { $messageCache = array(); }
        array_push($messageCache, $message);
        $message = $errors->first('.$expression.'); ?>';
    
    ...
    
    // For closing one
    return '<?php unset($message);
        if (! empty($messageCache)) { $message = array_pop($messageCache); }
```

But make sure to cover it with separate test if you do decide to get rid of this problem and introduce nesting support without potential side-effects.

**Comment by taylorotwell** [comment]
Maybe a paranoid concern but should we be concerned about merging this into 5.8 and stomping on someone's "macros"? Do macros take precedence over Blade defined directives?

**Comment by driesvints** [comment]
@taylorotwell I don't believe we need to take into account whatever anyone has added in user land. If that was the case than you couldn't add anymore new methods to classes etc because you'd "risk" it that someone else would have implemented that method. New functionality which doesn't breaks the public api should always be ok.

**Comment by rellect** [comment]
@driesvints I believe the concern is if to merge into 5.8 or master

**Comment by driesvints** [comment]
@rellect yes, that's what I was addressing.

**Comment by devcircus** [comment]
Usually I'd agree that we shouldn't waste too much time worrying about userland conflicts with custom macros/helpers, however I believe "error" is so common that we should consider sending to master and noting it in the upgrade guide. I have to assume there are many apps using an "error" helper.  I don't think this is such a pressing matter that we should risk breaking these apps on a point release. 

I'm not even sure which would take precedence, so it may not be an issue.

**Comment by driesvints** [comment]
@devcircus yeah, that's a good point actually.

**Comment by timacdonald** [comment]
I just confirmed that a user land defined directive takes precedence over a core directive.

```php
public function boot()
{
    Blade::directive('php', function ($expression) {
        return "<script>alert('You wanted PHP. You got JS. You\'re welcome.');</script>";
    });

    Blade::directive('csrf', function ($expression) {
        return "<script>alert('CSRF');</script>";
    });
}
```

```
<!-- welcome.blade.php -->
@php
@csrf
```

Screen Shot 2019-04-16 at 1 33 39 pm

Screen Shot 2019-04-16 at 1 33 41 pm

**Comment by devcircus** [comment]
If that's the case then I'm 👍🏻

**Review by akiyamaSM — APPROVED** [review]

**Comment by devcircus** [comment]
🎉🍾🍻🎉

**Comment by jsiebach** [comment]
This broke our app in a few places!  We have a lot of vue directives that use @error, ie.

    <vue-table @error="isError = true">

I have just changed that to 

    <vue-table {{'@'}}error="isError = true">

For now... Is there a better way to escape the @ symbol?

**Comment by timacdonald** [comment]
The quickest fix would probably be to just declare a custom directive.

```
public function boot()
{
    Blade::directive('error', function ($expression) {
        return "@error";
    });
}
```

that will restore your expected behaviour.

**Comment by jsiebach** [comment]
@timacdonald thanks. I found something saying `@@error` would work to escape it, and it looks like tests are passing.

I would think this would be a breaking change given how many Laravel devs must use `@error` in Vue... but for now I'm going to implement the `@@`

**Comment by driesvints** [comment]
@jsiebach we can't account for every X language out there. Since nothing changed to the public API of Laravel this wasn't a breaking change.

**Comment by jamesgraham** [comment]
Just done the upgrade and ran into this exact problem 👎
